### PR TITLE
[vitest-pool-workers] Pages Functions support, RPC support, E2E tests, and other fixes

### LINKED ISSUE #5288: 🚀 Feature Request: Example for Pages Functions using `@cloudflare/vitest-pool-workers`
### Describe the solution

For Pages Functions, the exported function requires `EventContext`, which is different from what's accepted by Workers, so `createExecutionContext ` can't be used. What's the best way to create such a context?

### LINKED ISSUE #5386: Ensure JSRPC/named entrypoints are supported for `SELF` binding and Durable Object bindings to same worker in Vitest integration
To support integration testing with hot-module-reloading, the Vitest integration wraps `export default` and any user-defined Durable Object classes. This wrapping will need to be updated for JSRPC. In particular, we need to make sure RPC methods can still be called on `WorkerEntrypoint` `default` exports, and `DurableObject` subclasses. Only prototype methods can be called over RPC, so we'll need to be think about how we expose methods from a dynamically `import()`ed class on the prototype of a class. For `DurableObject`s, we can likely do this by modifying the prototype in a `blockConcurrencyWhile()` callback. For `WorkerEntrypoint`s, we might have to do something super hacky with a `Proxy` prototype (I assume this is allowed 🙈)?

### LINKED ISSUE #5392: Additional meta-testing for the Workers Vitest integration
Currently, the Vitest integration is tested against a suite of example tests. There are some things that can't be tested this way though:

- [x] (Inline) snapshots, updating existing and remove obsolete
- [x] `console.log()`ing (including inside `SELF`) with source maps
- [x] Watch mode (for unit tests and `SELF` integration tests https://github.com/cloudflare/workers-sdk/issues/5384)
- [x] Config validation and pretty-formatting of validation errors
- [x] Concurrent tests failing with isolated storage enabled
- [x] Ensuring required compatibility flags set
- [x] Service worker `main` detection and failure
- [ ] Dynamic `import()` (for unit tests and `SELF` integration tests https://github.com/cloudflare/workers-sdk/issues/5387)
- [x] Large modules chunking

These require us to (re)start Vitest programatically and assert on its output, or interact with a running session. We'll also want to make sure we're testing...

- [x] Fake timers
- [x] Different isolation models
- [x] Mock fallthrough to the internet

### PR DESCRIPTION
**Pull request by mrbbot**
## What this PR solves / how to test

This PR implements a bunch of improvements to the Workers Vitest integration:

### Features

- Adds support for testing Pages Functions with helpers for creating `ASSETS` service bindings in config, and `EventContext`s in tests (fixes #5288)
- Adds support for RPC via `SELF` and named entrypoints (fixes #5386)
- Adds E2E testing for automatically-rerunning on changes, snapshots, console logging with source maps, etc (fixes #5392)

### Fixes

- #5384
- #5385
- #5268
- #5389
- #5390

## Author has addressed the following

- Tests
  - [ ] TODO (before merge)
  - [x] Included
  - [ ] Not necessary because:
- Changeset ([Changeset guidelines](https://github.com/cloudflare/workers-sdk/blob/main/CONTRIBUTING.md#changesets))
  - [ ] TODO (before merge)
  - [x] Included
  - [ ] Not necessary because:
- Public documentation
  - [x] TODO (before merge)
    - [ ] Document `buildPagesASSETSBinding()` here: https://developers.cloudflare.com/workers/testing/vitest-integration/configuration/#functions
    - [ ] Document `createPagesEventContext()` here: https://developers.cloudflare.com/workers/testing/vitest-integration/test-apis/#cloudflaretest-module-definition
    - [ ] Remove `Automatically re-running tests with SELF` and `console.log()s with SELF` from known issues here https://developers.cloudflare.com/workers/testing/vitest-integration/known-issues/
  - [ ] Cloudflare docs PR(s): <https://github.com/cloudflare/cloudflare-docs/pull/>...
  - [ ] Not necessary because:

### CODE REVIEW

**Comment by punkeel** [comment]
Is this the right issue to watch for Durable Objects RPC support in vitest? 👀

**Comment by punkeel** [comment]
Thanks for this PR, I can confirm RPC tests _almost work_ now! Looking forward to this hitting main!

I found a quirky bug: this only works well for simple types, not for objects.
Eg: 
```ts
export class Lala extends DurableObject {
    async ping() {
       return { x:42 };
    }
}
```

with the following test:
```ts
import { env } from 'cloudflare:test';
import { beforeEach, describe, it, expect } from 'vitest';

describe('checks', () => {
    it('bugs', async () => {
        const id = env.LALA.newUniqueId();
        const stub = env.LALA.get(id);
        const resp = await stub.ping();
        expect(resp.x).toBe(42);
    });
});
```

fails:

```console
npx vitest --run

 RUN  v1.3.0 /Users/maxime/Code/NC/workers-repro

[vpw:inf] Starting isolated runtimes for vitest.config.js...
[mf:wrn] The latest compatibility date supported by the installed Cloudflare Workers Runtime is "2024-04-03",
but you've requested "2024-04-05". Falling back to "2024-04-03"...

================================================================================
Failed to pop isolated storage stack frame in index.test.ts's test "bugs".
This usually means your Worker tried to access storage outside of a test.
In particular, we were unable to pop Durable Objects storage.
Ensure you `await` all `Promise`s that read or write to these services.

- AssertionError [ERR_ASSERTION]: Expected .sqlite, got /var/folders/47/5fsq322s5dx9y0f1c2shbvvh0000gn/T/miniflare-75b78b3ad39c574af773d9c1e217d656/do/vitest-pool-workers-runner--__VITEST_POOL_WORKERS_USER_OBJECTLala/f80a936e67292d7e11ad523172ca483114a78a52a221bec474662c3d8359ba6c.sqlite-shm
================================================================================

Error running worker: AssertionError [ERR_ASSERTION]: Isolated storage failed. There should be additional logs above.
    at new AssertionError (node-internal:internal_assertionerror:419:15)
    at Function.strictEqual (node-internal:internal_assert:200:11)
    at WorkersTestRunner.updateStackedStorage (Users/maxime/Code/NC/workers-repro/node_modules/@cloudflare/vitest-pool-workers/dist/worker/lib/cloudflare/test-runner.mjs:129:12)
    at WorkersTestRunner.onAfterRunSuite (Users/maxime/Code/NC/workers-repro/node_modules/@cloudflare/vitest-pool-workers/dist/worker/lib/cloudflare/test-runner.mjs:164:5)
    at runSuite (Users/maxime/Code/NC/workers-repro/node_modules/@vitest/runner/dist/index.js:875:5)
    at runFiles (Users/maxime/Code/NC/workers-repro/node_modules/@vitest/runner/dist/index.js:896:5)
    at startTests (Users/maxime/Code/NC/workers-repro/node_modules/@vitest/runner/dist/index.js:905:3)
    at Users/maxime/Code/NC/workers-repro/node_modules/vitest/dist/chunks/runtime-runBaseTests.0UwIvo_U.js:114:7
    at withEnv (Users/maxime/Code/NC/workers-repro/node_modules/vitest/dist/chunks/runtime-runBaseTests.0UwIvo_U.js:82:5)
    at run (Users/maxime/Code/NC/workers-repro/node_modules/vitest/dist/chunks/runtime-runBaseTests.0UwIvo_U.js:103:3)
 · index.test.ts (1)
   · checks (1)
     · bugs

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Unhandled Errors ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯

Vitest caught 1 unhandled error during the test run.
This might cause false positive tests. Resolve unhandled errors to make sure your tests are not affected.

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Unhandled Error ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
Error: Isolated storage failed. There should be additional logs above.
 ❯ new AssertionError node-internal:internal_assertionerror:419:15
 ❯ Function.strictEqual node-internal:internal_assert:200:11
 ❯ WorkersTestRunner.updateStackedStorage Users/maxime/Code/NC/workers-repro/node_modules/@cloudflare/vitest-pool-workers/dist/worker/lib/cloudflare/test-runner.mjs:129:12
 ❯ WorkersTestRunner.onAfterRunSuite Users/maxime/Code/NC/workers-repro/node_modules/@cloudflare/vitest-pool-workers/dist/worker/lib/cloudflare/test-runner.mjs:164:5
 ❯ runSuite Users/maxime/Code/NC/workers-repro/node_modules/@vitest/runner/dist/index.js:875:5
 ❯ runFiles Users/maxime/Code/NC/workers-repro/node_modules/@vitest/runner/dist/index.js:896:5
 ❯ startTests Users/maxime/Code/NC/workers-repro/node_modules/@vitest/runner/dist/index.js:905:3
 ❯ Users/maxime/Code/NC/workers-repro/node_modules/vitest/dist/chunks/runtime-runBaseTests.0UwIvo_U.js:114:7
 ❯ withEnv Users/maxime/Code/NC/workers-repro/node_modules/vitest/dist/chunks/runtime-runBaseTests.0UwIvo_U.js:82:5
 ❯ run Users/maxime/Code/NC/workers-repro/node_modules/vitest/dist/chunks/runtime-runBaseTests.0UwIvo_U.js:103:3

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯

 Test Files   (1)
      Tests   (1)
     Errors  1 error
   Start at  00:48:09
   Duration  762ms (transform 18ms, setup 0ms, collect 20ms, tests 0ms, environment 0ms, prepare 180ms)

[vpw:dbg] Shutting down runtimes...
make: *** [all] Error 1
```

[repro repo](https://github.com/punkeel/repro-5629/tree/maxime/5508)

This works with simple types, and seems to work if the RPC call is wrapped inside `runInDurableObject`.

**Comment by Cherry** [comment]
It looks like this might require a bump to Node 18, both for the minimal node version, and Actions. Vitest's officially minimum supported Node version is 18 anyway, so it was really just lucky that it was working up to this point.

There seems to be a comment in the code about this too:
```
// TODO(now): make Node 18 the minimum supported version
```

**Comment by punkeel** [comment]
https://github.com/vitest-dev/vitest/blob/v1.3.0/packages/vitest/package.json#L110-L112

vitest already enforces it, that should be a no brainer.

File: packages/wrangler/src/cli.ts
Comment (RamIdeas) [inline packages/wrangler/src/cli.ts:55]: This is a concern. I'm not sure of a better way of doing this right now but we likely want to find one. I guess we'd need to decouple the `@cloudflare/pages-shared/environment-polyfills/types.ts` `global` augmentations from this particular function.

Not a blocker as it's marked `unstable_` but highlighting.

File: fixtures/vitest-pool-workers-examples/rpc/vitest.config.ts
Comment (admah) [reply fixtures/vitest-pool-workers-examples/rpc/vitest.config.ts:11]: Yes, we need to. We need to document `SELF.scheduled()` as a whole, since it is missing from docs AFAICT.

**Comment by admah** [comment]
@punkeel yes, can you file a separate issue for that for us to track? Thanks!

**Comment by RamIdeas** [comment]
I've rebased from main to fix the conflicts

**Review by RamIdeas — APPROVED** [review]

**Comment by punkeel** [comment]
@RamIdeas can we please fix or document https://github.com/cloudflare/workers-sdk/pull/5508#issuecomment-2060035186 before merging? I think it should be a blocker 🙏

**Comment by RamIdeas** [comment]
@punkeel I spent some time today investigating the bug:

- returning primitives from RPC methods works fine
- returning plain objects from RPC methods fails
  - actually only from Durable Object RPC methods
  - returning objects from RpcEntrypoint methods works fine

```ts
// PASSES
it("can return POJOs from RpcTarget", async () => {
	const result = (await SELF.createCounter(5)).foo();
	expect(result).toEqual({ x: 42 });
});

// FAILS
it("can return POJOs from DO", async () => {
	const id = env.TEST_OBJECT.newUniqueId();
	const stub = env.TEST_OBJECT.get(id);
	const result = await stub.foo();

	expect(result).toEqual({ x: 42 });
});
```
- removing [this assertion](https://github.com/cloudflare/workers-sdk/blob/bcoll/vitest-integration-improvements/packages/vitest-pool-workers/src/pool/loopback.ts#L249) makes both tests pass
- BUT that will break isolated storage guarantees 
- I think the correct fix is to somehow earlier wait for the unfinished process of committing the .sqlite-shm file to a .sqlite file

I will continue to investigate tomorrow

**Comment by punkeel** [comment]
Interesting!
Any chance my assumption was wrong? I assumed RPC calls on DO should work without `runInDurableObject`, because it worked with primitive types. Was that assumption wrong, and the real "bug" here is that RPC call shouldn't work _at all_ without `runInDurableObject`?

(Does `fetch` work because it runs `runInDurableObject` internally, and RPC calls should but don't?)

I'm happy to test & provide feedback but I'm really out of my league here :)

**Comment by Skye-31** [comment]
There should not be a requirement on `runInDurableObject` - there's just a bug with returning non primitives from DO RPC at the moment

**Comment by ericmatthys** [comment]
It's been a few weeks of inactivity for this PR and it addresses quite a lot that I'm needing to wait for. Could you share an update? Or if it's blocked on something else that I could subscribe to, I'd be very grateful. 🙏

**Comment by petebacondarwin** [comment]
Hi @ericmatthys - the others have been moved onto other project work and I will take on this PR. I expect to start landing at least pieces of it early next week. Thanks for your patience.

**Comment by penalosa** [comment]
This PR is being landed in stages rather than all at once. All but https://github.com/cloudflare/workers-sdk/pull/5913 has been landed into main.

**Pull request closed without merging**
